src/qililab/analog/annealing_program.py

```diff
@@ -0,0 +1,93 @@
+
+
+class AnnealingProgram:
+    """Class for an Annealing Program. The program should have the format
+
+    .. code-block:: python
+
+        [
+            {"qubit_0": {"sigma_x" : 0, "sigma_y" : 1, "sigma_z" : 2},
+            "coupler_1_0 : {...},
+            },      # time=0ns
+            {...},  # time=1ns
+        .
+        .
+        .
+        ]
+
+
+    Args:
+        platform (Any): platform
+        annealing_program (list[dict[str, dict[str, float]]]): dictionary with the annealing program with the above structure.
+    """
+
+    def __init__(self, platform: Any, annealing_program: list[dict[str, dict[str, float]]]):
+        """Init method"""
+        self._platform = platform
+        self._annealing_program = annealing_program
+        self.annealing_program = annealing_program  # TODO: implement as frozenDataclass
+
+    def transpile(self, transpiler: Callable):
+        """First implementation of a transpiler, pretty basic but good as a first step. Transpiles from ising coefficients to fluxes
+
+        Args:
+            transpiler (Callable): Transpiler to use. The transpiler should take 2 values as arguments (delta, epsilon)
+            and return 2 values (phix, phiz)
+        """
+
+        # iterate over each anneal step and transpile ising to fluxes
+        for annealing_step in self._annealing_program:
+            for circuit_element in annealing_step:
+                phix, phiz = transpiler(
+                    delta=annealing_step[circuit_element]["sigma_x"], epsilon=annealing_step[circuit_element]["sigma_z"]
+                )
+                annealing_step[circuit_element]["phix"] = phix
+                annealing_step[circuit_element]["phiz"] = phiz
+
+    def get_waveforms(self) -> dict[str, tuple[Bus, ArbitraryWave]]:
+        """Returns a dictionary containing (bus, waveform) for each flux control from the transpiled fluxes. `AnnealingProgram.transpile` should be run first. The waveform is an arbitrary waveform obtained from the transpiled fluxes.
+
+        Returns:
+            dict[str,tuple[Bus,ArbitraryWave]]: Dictionary containing (bus, waveform) for each flux control (i.e. phix or phiz).
+        """
+        # parse names from algorithm notation (e.g. qubit_0) to runcard notation (e.g. q0)
+        element_name_map = {"qubit": "q", "coupler": "c"}
+        circuit_element_map = {
+            (element, flux): f"{flux}_{element_name_map[element.split('_', 1)[0]]}{element.split('_', 1)[1]}"
+            for element in self._annealing_program[0].keys()
+            for flux in self._annealing_program[0][element].keys()
+            if "phi" in flux
+        }  # {(element, flux): flux_line}
+
+        # Initialize dictionary with flux_lines pointing to (corresponding bus, waveform)
+        annealing_waveforms = {  # type: ignore[var-annotated]
+            flux_line: (self._platform.get_element(flux_line), []) for flux_line in circuit_element_map.values()
```

Comment:
Another important thing I just saw: Since we just need the Platform to retrieve the mapping of flux lines, we should avoid having this dependency.

Instead the parameter at the `__init__` method should be this mapping.

Also, if the `transpile()` method moves to the `Transpiler` class this dependency moves away from the `AnnealingProgram` completely.

Which makes sense, since our Programs (circuit / qprogram / annealing) are just representations and the compilation / execution are handled by other classes.

Reply:
There are 2 different transpilers here (technically). One is the one that handles ising to fluxes (this is the one I call transpiler), the other one handles fluxes to amplitudes (waveforms). We should ensure modularity of these 2 since they belong to 2 different workflows, even if they are both supposed to get data from calibration (but they should absolutely be run independently).
The `get_element` method is the standard method to get information for platform objects. Getting the mapping as an object would be better but that would require parsing the class input as a previous step in order to get the `{flux_control_line : bus}` dictionary (note that the current input expects `{chip_element : {ising_coeff : value}}`
This is related to https://github.com/qilimanjaro-tech/qililab/pull/767#discussion_r1713813800 probably; again we can standarize the input once we have more information about the upper stack, I don't see the point in doing this now

Reply:
> There are 2 different transpilers here (technically). One is the one that handles ising to fluxes (this is the one I call transpiler), the other one handles fluxes to amplitudes (waveforms). We should ensure modularity of these 2 since they belong to 2 different workflows, even if they are both supposed to get data from calibration (but they should absolutely be run independently).

This makes me want the usage of classes even more!

> The `get_element` method is the standard method to get information for platform objects. Getting the mapping as an object would be better but that would require parsing the class input as a previous step in order to get the `{flux_control_line : bus}` dictionary.

Yes, I am not against using `get_element()`, this is the correct way indeed. I am against having a dependency on platform to build an `AnnealingProgram` instance. The annealing program should be independent of the platform. If it needs a mapping `{flux_control_line : bus}` information then this should be the parameter of the constructor, not the whole platform.

There is a similar situation in `QbloxCompiler` where we need the `time_of_flight` for each bus, which we get from platform. The signature of the `compile()` method expects just that: `times_of_flight: dict[str, int] | None = None`. The user might create the dictionary manually (providing total flexibility) or in the case that you execute qprogram from `platform.execute_qprogram`, we retrieve the mapping from platform:

```
times_of_flight = {
                bus.alias: int(bus.get_parameter(Parameter.TIME_OF_FLIGHT))
                for bus in buses
                if isinstance(bus.system_control, ReadoutSystemControl)
            }
```